### PR TITLE
Add exec option to tmpfs

### LINKED ISSUE #32131: Swarm mode, mount tmpfs, how to remove noexec?
**Description**

Swarm mode, ```docker service create --mount type=tmpfs```,  how to remove noexec flag?

**Steps to reproduce the issue:**
1.  docker service create --mount type=tmpfs,destination=/data,tmpfs-size=100M -p:22 rastasheep/ubuntu-sshd:14.04
2. ssh root@localhost -p 30000
3. mount | grep /data

**Describe the results you received:**
tmpfs on /data type tmpfs (rw,nosuid,nodev,noexec,relatime,size=102400k)

**Describe the results you expected:**
I want to remove noexec flag, how? like, ```docker run -it --rm --tmpfs /data:exec bash```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:53:04 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:53:04 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 7
Server Version: 17.03.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 72
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: 8t1tc9olkkkvto0m9ap20yg4q
 Is Manager: true
 ClusterID: mjf13s8g3luwoo579kbtkxvau
 Managers: 1
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.25.20.22
 Manager Addresses:
  172.25.20.22:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Security Options:
 apparmor
Kernel Version: 3.13.0-98-generic
Operating System: Ubuntu precise (12.04.5 LTS)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 993.5 MiB
Name: docker0
ID: HPTL:DPNP:QNJ6:6K5R:NFNT:CMFE:RWWP:TWAW:5DAO:LZXV:DEKY:5VKZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 44
 Goroutines: 136
 System Time: 2017-03-27T10:03:52.81516827Z
 EventsListeners: 1
Username: huozic
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.): **
VirtualBox 5.1.16 r113841
ubuntu 12.04 3.13.0-98-generic

### LINKED ISSUE #45385: Add `--mount` tmpfs with `exec` support for vscode devcontainer
### Description

vscode devcontainer.json
```
  "mounts": [
    "type=tmpfs,destination=/tmp,tmpfs-mode=1777,exec=true",
```

```
invalid argument "type=tmpfs,destination=/tmp,tmpfs-mode=1777,exec=true" for "--mount" flag: unexpected key 'exec' in 'exec=true'
See 'docker run --help'.
```

---

vscode devcontainer.json
```
  "mounts": [
    "type=tmpfs,destination=/tmp,tmpfs-mode=1777,exec",
```

```
invalid argument "type=tmpfs,destination=/tmp,tmpfs-mode=1777,exec" for "--mount" flag: invalid field 'exec' must be a key=value pair
See 'docker run --help'.
```

---

https://github.com/moby/moby/issues/32131#issuecomment-1031359653

https://github.com/moby/sys/pull/66

That pr is merged, but what happend here?

### PR DESCRIPTION
**Pull request by adshmh**
This PR adds the ability to handle exec option on tmpfs mounts. 

- fixes #32131 
- fixes https://github.com/moby/moby/issues/45385

Notes:

This is the first draft. As discussed here: https://github.com/moby/moby/issues/32131#issuecomment-375074893, I have included the changes on `docker/swarmkit` types that will need vendoring. I will submit PR(s) on `docker/swarmkit` and `docker/cli` once if this PR is approved.

**- What I did**
Feature: allow specifying exec option for tmpfs mounts

**- How I did it**
Added the required code and tests to pkg/mount, volume, api/types, and daemon.

**- How to verify it**
I have added an integration test for verification. 

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by adshmh** [comment]
I think the integration test in `integration/container/mounts_linux_test.go` would read better using assertions, but followed the existing style on the file.

I can add a commit to use assertions for all the tests in `integration/container/mounts_linux_test.go` if that is a better option.

**Comment by justincormack** [comment]
needs a rebase

**Comment by cpuguy83** [comment]
Is changing tmpfs to use noexec by default breaking?

File: api/types/mount/mount.go
Comment (thaJeztah) [inline api/types/mount/mount.go]: Wondering if this needs a dedicated option, or if we should "punch through" mount options. Bit on the fence for that; pro is that we separate the API from the underlying implementation, but con would be that we may end up having an API change for every mount option that exists.

**Comment by kolyshkin** [comment]
I concur @thaJeztah idea to have a more generic way to pass options to tmpfs rather than implementing a specific option for exec. The options specified should be validated against a whitelist.

We have to figure which options to allow but let's at least include `exec` and `noexec`, but maybe others that are valid for `tmpfs` (see `man 8 mount` and the comment in https://github.com/moby/moby/blob/8bb5a28eed5eba5651c6e48eb401c03be938b4c1/api/types/mount/mount.go#L112).

We can add `nr_inodes` and `mpol` to the whitelist.

Options such as `uid` and `gid` might not work well in a cluster environment so they should probably not be added at this time.

Options `nr_blocks` and `size` should not be added, as they are already handled.

@adshmh would you wish to work on that?

**Comment by adshmh** [comment]
@kolyshkin Sure, I can update the PR. 

Is the following correct regarding the requirements?

- Allow a more generic way of passing options to tmpfs (perhaps a string)
- A white list for the allowed options
- Currently allow: `exec`, `noexec`, `nr_inodes` and `mpol`

File: api/types/mount/mount.go
Comment (adshmh) [reply api/types/mount/mount.go]: Thank you for the review. I can change that to allow `exec` and any other required options in a generic option passed to Tmpfs mount types.

**Comment by thaJeztah** [comment]
> Allow a more generic way of passing options to tmpfs (perhaps a string)

From a CLI /UX perspective, this could use the same as `docker volume create --opt foo=bar --opt bar=baz`

> A white list for the allowed options

Yes; we discussed this, and generally it's easier to _add_ more options later than _removing_ an option (as that would be a breaking change), so for that reason, we prefer a whitelist.

> `nr_inodes` and `mpol`

I'm not _against_ adding them, just wondering how common they are (i.e. should we add those now, or add them once we see a need); @kolyshkin ?

**Comment by thaJeztah** [comment]
Discussing in the maintainers meeting with @kolyshkin @cpuguy83, and having these options as a string should be ok for now; no reason to make it more complicated. Moving this to code review

**Comment by thaJeztah** [comment]
This also needs a PR in swarmkit to make the required changes there

**Comment by adshmh** [comment]
Thank you for the review. I will submit PRs on swarmkit and docker/cli.

**Comment by adshmh** [comment]
Rebased the PR to resolve the conflicts.

**Comment by dperny** [comment]
It looks to me, from the `tmpfs(5)` man page, that tmpfs only has a handful of options, which we could easily implement as discrete types. This gives us more control over tmfps options later on, so that we could theoretically do things like filter tmpfs options or enforce security policies in a swarm. But, doing a bit more digging, it seems that some options, like `exec`, come from `mount` and not `tmpfs`? Which means that raw tmpfs options isn't _really_ what this is, it's more like raw mount options, no?

**Comment by dperny** [comment]
Is the `exec`/`noexec` option specific to tmpfs? I'm unclear why tmpfs is special in this regard as opposed to other mount types.

**Comment by adshmh** [comment]
@dperny Thank you for the review. 

I think the `noexec` option is set as default only on mounts of type tmpfs, so the PR scope was limited to allowing an override of this.

The result of the first round of reviews was that using a general options field, validated against a whitelist, is preferable. If a boolean `exec` field, or a boolean field on `swarmkit` side, is need, please let me know and I'll update the PR accordingly.

File: pkg/mount/mount_unix_test.go
Comment (kolyshkin) [inline pkg/mount/mount_unix_test.go]: Use `GetMounts(SingleEntryFilter(target))` here, it will return a list of 0 or 1 elements so you don't have to loop through those.

Something like
```go
mounts, err := GetMounts(SingleEntryFilter(target))
if err != nil {
        t.Fatal(err)
}
if len(mounts) != 1 {
        t.Fatal("Mount point ", target, " not found")
}
entry := mounts[0]
if !strings.Contains(entry.Opts, tc.expected) {
        t.Fatal("Expected option ", tc.expected, " missing from ", entry.Opts)
}
if tc.unexpected != "" && strings.Contains(entry.Opts, tc.unexpected) {
        t.Fatal("Unexpected mount option ", tc.unexpected, " in ", entry.Opts)
}
```

File: pkg/mount/mount_unix_test.go
Comment (kolyshkin) [inline pkg/mount/mount_unix_test.go]: maybe s/expectedFlag/expectedFlags/ while you're at it.

File: pkg/mount/mount_unix_test.go
Comment (kolyshkin) [inline pkg/mount/mount_unix_test.go]: This is not very important here, but the options are comma-separated list of entries and should be treated as such, so substring search is ambiguous here.

I'd suggest having something like
```go
opts := "," + entry.Opts + ","
if !strings.Contains(opts, ","+tc.expected+",") {
...
```
or something like this. This way you can add "exec" to "noexpected" for the second test case.

File: api/types/mount/mount.go
Comment (kolyshkin) [reply api/types/mount/mount.go]: Maybe we can just name this `Options`, this is in-line with the language used around mounts?

File: daemon/oci_linux_test.go
Comment (kolyshkin) [inline daemon/oci_linux_test.go]: I don't see how this test case brings some additional value. Checking that exec and noexec can be used and result in mounts with proper options is already done in a previous commit. I would drop this one.

File: integration/container/mounts_linux_test.go
Comment (kolyshkin) [inline integration/container/mounts_linux_test.go]: Same here -- I would drop this test case, I don't see how it can bring some additional value.

If you don't want to drop it, at least use `container.Exec()` from `integration/internal/container/exec.go`, it will greatly simplify the code.

**Review by kolyshkin — CHANGES_REQUESTED** [review]

**Comment by kolyshkin** [comment]
I would vote against a boolean field. A comma-separated list is simple, extensible, and easy to validate.

Yes, `noexec` option is not specific to tmpfs, but we have tmpfs as a distinct fs type so...

File: pkg/mount/mount_unix_test.go
Comment (adshmh) [reply pkg/mount/mount_unix_test.go]: Thank you for the detailed review. I will update the PR, sorry for the delay.

File: pkg/mount/mount_unix_test.go
Comment (adshmh) [reply pkg/mount/mount_unix_test.go]: Thank you, fixed.

File: pkg/mount/mount_unix_test.go
Comment (adshmh) [reply pkg/mount/mount_unix_test.go]: Thanks, yes this will allow to add the `exec` option to `unexpected` field. Updated the PR.

File: daemon/oci_linux_test.go
Comment (adshmh) [reply daemon/oci_linux_test.go]: Thanks, removed. I had added it to test for the default value of `noexec` set by the daemon on tmpfs volumes, but that default value is not introduced by this PR anyway.

File: integration/container/mounts_linux_test.go
Comment (adshmh) [reply integration/container/mounts_linux_test.go]: Thank you, removed.

**Comment by thaJeztah** [comment]
> I would vote against a boolean field. A comma-separated list is simple, extensible, and easy to validate.
> Yes, noexec option is not specific to tmpfs, but we have tmpfs as a distinct fs type so...

I'm with @kolyshkin on this; we discussed some options in a maintainers meeting, and after discussing alternatives, we concluded that we should prevent over-engineering/abstracting this, and keep things simple.

@dperny is this a big blocker for you / the swarmkit maintainers?

**Comment by AkihiroSuda** [comment]
What's current status?

**Comment by thaJeztah** [comment]
Asked @dperny again on the related SwarmKit PR

**Comment by denismakogon** [comment]
2 month from the last comment, what's the status? has to be fixed soon, since this is an actual problem to certain types of workload.

**Comment by thaJeztah** [comment]
@denismakogon such comments are not very helpful.

This PR depends on https://github.com/docker/swarmkit/pull/2647 to be accepted and merged in SwarmKit first; after that, work on this PR can continue.

**Comment by denismakogon** [comment]
@thaJeztah well, maybe, however, the progress is kinda slow even having all PRs ready and the design was approved. So, the pace of getting things in is not quite impressive, but I guess that's what it takes.

Any estimates so far?

**Comment by denismakogon** [comment]
boom, merged: https://github.com/docker/swarmkit/pull/2647

**Comment by denismakogon** [comment]
@adshmh can you please update the PR?

**Comment by adshmh** [comment]
I think the next step is to vendor the changes that are merged in swarmkit. I will submit a PR for this shortly.

**Comment by denismakogon** [comment]
@adshmh thanks, a lot.

**Comment by denismakogon** [comment]
@adshmh hi, were you able to submit necessary changes?

**Comment by thaJeztah** [comment]
@denismakogon please be patient; you're generating a lot of noise. This feature involves an API change, so won't be available in a release until the next release (Docker 19.03) when merged, but you will be able to test it on the nightly builds before that.

**Comment by adshmh** [comment]
@thaJeztah I just noticed #38123 is open to vendor swarmkit. Should I submit a separate PR to vendor swarmkit changes related to tmpfs, or would you prefer to extend the existing PR?

**Comment by denismakogon** [comment]
@adshmh @thaJeztah hi guys, any progress on this PR?

**Comment by olljanat** [comment]
@adshmh this one needs to be rebased

**Comment by olljanat** [comment]
Derek add label: status/failing-ci

**Comment by adshmh** [comment]
@olljanat thanks for the reminder. I will rebase the PR.

**Comment by olljanat** [comment]
Derek add label: rebuild/z

**Comment by olljanat** [comment]
Derek remove label: status/failing-ci

**Comment by olljanat** [comment]
Derek add label: rebuild/z

File: api/types/mount/mount.go
Comment (thaJeztah) [inline api/types/mount/mount.go]: Can you add this to the Swagger.yml as well, and to the API history https://github.com/moby/moby/blob/b4842cfe88b39af42416429396aee10d1fe7c3d6/docs/api/version-history.md#v140-api-changes? https://github.com/moby/moby/blob/b4842cfe88b39af42416429396aee10d1fe7c3d6/api/swagger.yaml#L297-L307


@kolyshkin did we want this to be a `[]string` or `map[string]string` ?

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: api/types/mount/mount.go
Comment (adshmh) [reply api/types/mount/mount.go]: Thank you for the review. I updated the `swagger.yaml` and the `version-history.md` files, assuming `string` for the `Options` field as I recall from previous discussions. If a `map[string]string` is to be used instead, please let me know.

**Comment by cpuguy83** [comment]
What's the status on this?

**Comment by hmh** [comment]
Using docker for CI/CD that requires lots of unpacking, build, throw-away is about 10 times fast with *proper* use of tmpfs, and this almost always requires exec.

What is needed to get this thing out of the door? I mean, it has been what, two years that this is languishing ?

**Comment by AkihiroSuda** [comment]
At least this needs to be rebased

**Comment by AkihiroSuda** [comment]
Needs another rebase.
LGTM if CI green

**Comment by brunoais** [comment]
@adshmh Any updates?

**Comment by jehon** [comment]
Any news? Any help needed?

**Pull request closed without merging**
